test/integration/test_save_job_id_on_datasets.py

```diff
@@ -0,0 +1,40 @@
+"""
+A Dataset should have a reference to the Job that created it.
+Here we test the creation of this reference in various contexts:
+- job creates one dataset
+- job creates multiple datasets
+- job uploads a dataset
+- job creates dynamically discovered datasets (after the tool is run)
+"""
+
+import pytest
+
+from galaxy import model
+from galaxy_test.driver.driver_util import GalaxyTestDriver
+
+TEST_TOOL_IDS = [
+    'boolean_conditional',  # 1 job >> 1 dataset
+    'color_param',          # 1 job >> multiple datasets
+    'multi_output',         # 1 job >> dataset + discovered dataset? (maybe; not implemented yet)
+    # test upload?            # not implemented
+]
+
+
+class TestConfig:
+    framework_tool_and_types = True
+
+
+@pytest.mark.parametrize('tool_id', TEST_TOOL_IDS)
+def test_tool_datasets(tool_id):
+    driver = GalaxyTestDriver()
+    driver.setup(TestConfig)
+    driver.run_tool_test(tool_id)
+    session = driver.app.model.context.current
+    job = session.query(model.Job).one()
```

Comment:
The database is not created for this one test, so this won't work: there are multiple jobs - hence the error.